File: app/src/main/java/com/fatbook/fatbookapp/ui/activity/SignInActivity.java
package com.fatbook.fatbookapp.ui.activity;

import android.content.Intent;
import android.graphics.Rect;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.view.ViewTreeObserver;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.content.res.AppCompatResources;
import androidx.core.content.ContextCompat;
import androidx.lifecycle.ViewModelProvider;

import com.fatbook.fatbookapp.R;
import com.fatbook.fatbookapp.core.Ingredient;
import com.fatbook.fatbookapp.core.IngredientUnit;
import com.fatbook.fatbookapp.core.Recipe;
import com.fatbook.fatbookapp.core.RecipeIngredient;
import com.fatbook.fatbookapp.core.Role;
import com.fatbook.fatbookapp.core.User;
import com.fatbook.fatbookapp.databinding.ActivitySignInBinding;
import com.fatbook.fatbookapp.retrofit.RetrofitFactory;
import com.fatbook.fatbookapp.ui.viewmodel.SignInViewModel;
import com.fatbook.fatbookapp.util.UserUtils;

import java.util.ArrayList;
import java.util.List;
import java.util.logging.Level;

import lombok.extern.java.Log;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

@Log
public class SignInActivity extends AppCompatActivity {

    private ActivitySignInBinding binding;

    private SignInViewModel signInViewModel;

    public boolean isKeyboardVisible = false;

    public final ViewTreeObserver.OnGlobalLayoutListener listener = new ViewTreeObserver.OnGlobalLayoutListener() {
        @Override
        public void onGlobalLayout() {
            final Rect rectangle = new Rect();
            final View contentView = binding.getRoot();
            contentView.getWindowVisibleDisplayFrame(rectangle);
            int screenHeight = contentView.getRootView().getHeight();
            int keypadHeight = screenHeight - rectangle.bottom;
            boolean isKeyboardNowVisible = keypadHeight > screenHeight * 0.15;

            if (isKeyboardVisible != isKeyboardNowVisible) {
                if(isKeyboardNowVisible) {
                    binding.textViewSignInVersion.setVisibility(View.GONE);
                    binding.textViewSignInCopyright.setVisibility(View.GONE);
                    binding.textViewSignInTagline.setVisibility(View.GONE);
                    binding.textViewSignInAppLabel.setVisibility(View.GONE);
                } else {
                    binding.textViewSignInVersion.setVisibility(View.VISIBLE);
                    binding.textViewSignInCopyright.setVisibility(View.VISIBLE);
                    binding.textViewSignInTagline.setVisibility(View.VISIBLE);
                    binding.textViewSignInAppLabel.setVisibility(View.VISIBLE);
                }
            }
            isKeyboardVisible = isKeyboardNowVisible;
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivitySignInBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        signInViewModel = new ViewModelProvider(this).get(SignInViewModel.class);

        signInViewModel.getUser().observe(this, _user -> {
            if (_user != null) {
                Intent intent = new Intent(getApplicationContext(), MainActivity.class);
                intent.putExtra(UserUtils.TAG_USER, _user);
                startActivity(intent);
                finishAffinity();
            } else {
                binding.editTextSignInLogin.setBackground(AppCompatResources.getDrawable(this, R.drawable.round_corner_edittext_login_error));
                binding.editTextSignInPassword.setBackground(AppCompatResources.getDrawable(this, R.drawable.round_corner_edittext_login_error));
            }
        });

        binding.buttonSignIn.setOnClickListener(view -> {
            validateLogin(binding.editTextSignInLogin.getText().toString().trim(), binding.editTextSignInPassword.getText().toString());
        });

        binding.editTextSignInLogin.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                binding.editTextSignInLogin.setBackground(AppCompatResources.getDrawable(getBaseContext(), R.drawable.round_corner_edittext_login));
                binding.editTextSignInPassword.setBackground(AppCompatResources.getDrawable(getBaseContext(), R.drawable.round_corner_edittext_login));
            }

            @Override
            public void afterTextChanged(Editable editable) {
            }
        });

        binding.editTextSignInPassword.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                binding.editTextSignInLogin.setBackground(AppCompatResources.getDrawable(getBaseContext(), R.drawable.round_corner_edittext_login));
                binding.editTextSignInPassword.setBackground(AppCompatResources.getDrawable(getBaseContext(), R.drawable.round_corner_edittext_login));
            }

            @Override
            public void afterTextChanged(Editable editable) {
            }
        });
    }

    private void validateLogin(String login, String fat) {
       RetrofitFactory.apiServiceClient().signIn(login, fat).enqueue(new Callback<User>() {
            @Override
            public void onResponse(@NonNull Call<User> call, @NonNull Response<User> response) {
                log.log(Level.INFO, "fat: " + response.code());
                if (response.code() == 200) {
                    signInViewModel.setUser(response.body());
                } else {
                    signInViewModel.setUser(null);
                }
            }

            @Override
            public void onFailure(@NonNull Call<User> call, @NonNull Throwable t) {
                log.log(Level.INFO, "fat: FAILED " + t);
                signInViewModel.setUser(null);
            }
        });
    }

    @Override
    protected void onResume() {
        super.onResume();
        binding.getRoot().getViewTreeObserver().addOnGlobalLayoutListener(listener);
    }

    @Override
    protected void onPause() {
        super.onPause();
        binding.getRoot().getViewTreeObserver().removeOnGlobalLayoutListener(listener);
    }
}

File: app/src/main/java/online/fatbook/fatbookapp/ui/activity/SignInActivity.kt
package online.fatbook.fatbookapp.ui.activity

import android.content.Intent
import android.graphics.Rect
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.*
import android.view.ViewTreeObserver.OnGlobalLayoutListener
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.content.res.AppCompatResources
import androidx.lifecycle.ViewModelProvider
import lombok.extern.java.Log
import online.fatbook.fatbookapp.R
import online.fatbook.fatbookapp.core.User
import online.fatbook.fatbookapp.databinding.ActivitySignInBinding
import online.fatbook.fatbookapp.retrofit.RetrofitFactory
import online.fatbook.fatbookapp.ui.viewmodel.SignInViewModel
import online.fatbook.fatbookapp.util.UserUtils
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.logging.Level

class SignInActivity : AppCompatActivity() {
    private var binding: ActivitySignInBinding? = null
    private var signInViewModel: SignInViewModel? = null
    var isKeyboardVisible = false
    val listener = OnGlobalLayoutListener {
        val rectangle = Rect()
        val contentView: View = binding!!.root
        contentView.getWindowVisibleDisplayFrame(rectangle)
        val screenHeight = contentView.rootView.height
        val keypadHeight = screenHeight - rectangle.bottom
        val isKeyboardNowVisible = keypadHeight > screenHeight * 0.15
        if (isKeyboardVisible != isKeyboardNowVisible) {
            if (isKeyboardNowVisible) {
                binding!!.textViewSignInVersion.visibility = View.GONE
                binding!!.textViewSignInCopyright.visibility = View.GONE
                binding!!.textViewSignInTagline.visibility = View.GONE
                binding!!.textViewSignInAppLabel.visibility = View.GONE
            } else {
                binding!!.textViewSignInVersion.visibility = View.VISIBLE
                binding!!.textViewSignInCopyright.visibility = View.VISIBLE
                binding!!.textViewSignInTagline.visibility = View.VISIBLE
                binding!!.textViewSignInAppLabel.visibility = View.VISIBLE
            }
        }
        isKeyboardVisible = isKeyboardNowVisible
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivitySignInBinding.inflate(layoutInflater)
        setContentView(binding!!.root)
        signInViewModel = ViewModelProvider(this).get(SignInViewModel::class.java)
        signInViewModel!!.user.observe(this) { _user: User? ->
            if (_user != null) {
                val intent = Intent(applicationContext, MainActivity::class.java)
                intent.putExtra(UserUtils.TAG_USER, _user)
                startActivity(intent)
                finishAffinity()
            } else {
                binding!!.editTextSignInLogin.background = AppCompatResources.getDrawable(
                    this,
                    R.drawable.round_corner_edittext_login_error
                )
                binding!!.editTextSignInPassword.background = AppCompatResources.getDrawable(
                    this,
                    R.drawable.round_corner_edittext_login_error
                )
            }
        }
        binding!!.buttonSignIn.setOnClickListener { view: View? ->
            validateLogin(
                binding!!.editTextSignInLogin.text.toString().trim { it <= ' ' },
                binding!!.editTextSignInPassword.text.toString()
            )
        }
        binding!!.editTextSignInLogin.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                binding!!.editTextSignInLogin.background = AppCompatResources.getDrawable(
                    baseContext, R.drawable.round_corner_edittext_login
                )
                binding!!.editTextSignInPassword.background = AppCompatResources.getDrawable(
                    baseContext, R.drawable.round_corner_edittext_login
                )
            }

            override fun afterTextChanged(editable: Editable) {}
        })
        binding!!.editTextSignInPassword.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                binding!!.editTextSignInLogin.background = AppCompatResources.getDrawable(
                    baseContext, R.drawable.round_corner_edittext_login
                )
                binding!!.editTextSignInPassword.background = AppCompatResources.getDrawable(
                    baseContext, R.drawable.round_corner_edittext_login
                )
            }

            override fun afterTextChanged(editable: Editable) {}
        })
    }

    private fun validateLogin(login: String, fat: String) {
        RetrofitFactory.apiServiceClient().signIn(login, fat).enqueue(object : Callback<User?> {
            override fun onResponse(call: Call<User?>, response: Response<User?>) {
//                SignInActivity.log.log(Level.INFO, "fat: " + response.code())
                if (response.code() == 200) {
                    signInViewModel!!.user.value = response.body()
                } else {
                    signInViewModel!!.user.value = null
                }
            }

            override fun onFailure(call: Call<User?>, t: Throwable) {
//                SignInActivity.log.log(Level.INFO, "fat: FAILED $t")
                signInViewModel!!.user.value = null
            }
        })
    }

    override fun onResume() {
        super.onResume()
        binding!!.root.viewTreeObserver.addOnGlobalLayoutListener(listener)
    }

    override fun onPause() {
        super.onPause()
        binding!!.root.viewTreeObserver.removeOnGlobalLayoutListener(listener)
    }
}